### ISSUE TITLE
blank index的问题

### ISSUE THREAD

**Issue by yflv-yanxia** [thu-spmi/CAT#46]
虽然我已经反复的看了问题[#11](https://github.com/thu-spmi/CAT/issues/11)和问题[#22](https://github.com/thu-spmi/CAT/issues/22)的回答，但是依然对blank index的用法不太清楚，在此询问下。
训练时：
costs_ctc = torch.zeros(logits.size(0))
act = torch.transpose(logits, 0, 1).contiguous()
grad_ctc= torch.zeros(act.size()).type_as(logits)
ctc_crf_base.gpu_ctc(act, grad_ctc, labels, label_lengths, input_lengths, logits.size(0), costs_ctc, 0)
ctc_crf_base.gpu_den(logits, grad_den, input_lengths.cuda(), costs_alpha_den, costs_beta_den)
由于gpu_ctc函数的参数blank设置为0，所以logits中**blank index必然为0**，所以gpu_den的输入blank index也为0

生成denominator时：
run.sh中：
# Prepare denominator
    python3 ctc-crf/prep_ctc_trans.py data/lang_phn/lexicon_numbers.txt data/train_tr95/text "<UNK>" > data/train_tr95/text_number || exit 1
    cat data/train_tr95/text_number | sort -k 2 | uniq -f 1 > data/train_tr95/unique_text_number || exit 1
    mkdir -p data/den_meta
    chain-est-phone-lm ark:data/train_tr95/unique_text_number data/den_meta/phone_lm.fst || exit 1
    python3 ctc-crf/ctc_token_fst_corrected.py den data/lang_phn/tokens.txt | fstcompile | fstarcsort --sort_type=olabel > data/den_meta/T_den.fst || exit 1
    fstcompose data/den_meta/T_den.fst data/den_meta/phone_lm.fst > data/den_meta/den_lm.fst || exit 1
    echo "Prepare denominator finished"

我查看了lexicon.txt、units.txt和lexicon_num.txt，推理出unique_text_number 中的**blank index依然是0？**
虽然tokens.txt中<blk>为1，但是ctc_token_fst_corrected.py程序中
for line in lines:
    sp = line.split()
    phone = sp[0]
    if phone == '<eps>' or phone == '<blk>':
      continue
<eps>和<blk>都continue了，所以意味着**blank index依然是0吗？**
为什么在[#11](https://github.com/thu-spmi/CAT/issues/11)中，您解答blank index为1，这个地方不太懂。
因为我是做OCR方向的，所以需要搞懂这个问题，才可以正确应用到我的工程中去。

**Comment by hyx16** [comment]
首先总结一下：在我们的实现中，传入Loss function的神经网络输出都是“blank_index = 0"。（比如神经网络输出是netout这个tensor，netout的shape是(N, T, D)，那么netout[:, :, 0] 对应了blank的相应输入。
下面是具体问题：

> 由于gpu_ctc函数的参数blank设置为0，所以logits中blank index必然为0，所以gpu_den的输入blank index也为0

是这样的

> 我查看了lexicon.txt、units.txt和lexicon_num.txt，推理出unique_text_number 中的blank index依然是0？

可以这么理解

> \<eps\>和\<blk\>都continue了，所以意味着blank index依然是0吗？

input的blank index不是0，0是\<eps\>, 1是\<blk\>，ctc_token_fst_corrected.py代码中有一句def il(n): return n+1。”\<eps\>和\<blk\>都continue了“只是为了统计实际phone的数量。比如统计出实际的phone有10个，如果没有”n+1"，那么构图时的input label会是1, 2, 3...., 10。由于有"n+1"，input label变成了2, 3, ......, 11，留下来的0和1分别给\<eps\>和\<blk\>。

> 为什么在#11中，您解答blank index为1，这个地方不太懂。

这个其实想表达的是，我们构建好的用于计算的分母图denominator graph上，input label中\<eps\>是0，\<blk\>是1。
但是神经网络的输入还是“blank_index = 0"。我们在用于实际计算读入分母图时，做了处理又把边上的input label减1了，具体可见fst_read.cc里面，
`beta_ilabel[state].push_back(aiter.Value().ilabel-1);`
`beta_ilabel[state].push_back(aiter.Value().ilabel-1);`。

**Comment by yflv-yanxia** [comment]
多谢及时的解答
由于对fst不是很了解，再请教个问题：
wjs/run.sh中：
chain-est-phone-lm ark:data/train_tr95/unique_text_number data/den_meta/phone_lm.fst || exit 1
    python3 ctc-crf/ctc_token_fst_corrected.py den data/lang_phn/tokens.txt | fstcompile | fstarcsort --sort_type=olabel > data/den_meta/T_den.fst || exit 1
    fstcompose data/den_meta/T_den.fst data/den_meta/phone_lm.fst > data/den_meta/den_lm.fst || exit 1
unique_text_number文件中实际phone的index是从1开始的，\<NSN\>的index是1，所以由此生成的phone_lm.fst中的\<NSN\>的index是1，**还是说unique_text_number文件中的实际phone的index应该从2开始，\<NSN\>的index应该是2？**
由于\<eps\>是0，\<blk\>是1，由data/lang_phn/tokens.txt生成的T_den.fst中，\<NSN\>的index是2
由phone_lm.fst和T_den.fst共同生成的den_lm.fst
但是phone_lm.fst和T_den.fst中的实际phone的index不一致？

**Comment by hyx16** [comment]
> 还是说unique_text_number文件中的实际phone的index应该从2开始，\<NSN\>的index应该是2？

unique_text_number中没有\<blk\>，实际phone的index是从1开始

T_den.fst的input label中有\<blk\>，因此0是\<eps\>，1是\<blk\>，从2开始是实际phone，
output_label中没有\<blk\>，实际phone的index是从1开始

phone_lm.fst的input label和output label实际phone的index是从1开始。

然后T_den.fst和phone_lm.fst做compose后生成den_lm.fst，input label中有\<blk\>，从2开始是实际phone，
output_label中没有\<blk\>，实际phone的index是从1开始

den_lm中我们不关心output label，只关心input label。

**Comment by yflv-yanxia** [comment]
好的，明白了，多谢耐心解答
